synapse/_scripts/register_new_matrix_user.py

```diff
@@ -203,31 +205,74 @@ def main() -> None:
 
     parser.add_argument(
         "server_url",
-        default="https://localhost:8448",
         nargs="?",
-        help="URL to use to talk to the homeserver. Defaults to "
-        " 'https://localhost:8448'.",
+        help="URL to use to talk to the homeserver. By default, tries to find a "
+        "suitable URL from the configuration file. Otherwise, defaults to "
+        f"'{_DEFAULT_SERVER_URL}'.",
     )
 
     args = parser.parse_args()
 
     if "config" in args and args.config:
         config = yaml.safe_load(args.config)
+
+    if args.shared_secret:
+        secret = args.shared_secret
+    else:
+        # argparse should check that we have either config or shared secret
+        assert config
+
         secret = config.get("registration_shared_secret", None)
         if not secret:
             print("No 'registration_shared_secret' defined in config.")
             sys.exit(1)
+
+    if args.server_url:
+        server_url = args.server_url
+    elif config:
+        server_url = _find_client_listener(config)
+        if not server_url:
+            server_url = _DEFAULT_SERVER_URL
+            print(
+                "Unable to find a suitable HTTP listener in the configuration file. "
+                f"Trying {server_url} as a last resort.",
+                file=sys.stderr,
+            )
```

Comment:
Could we use `config.server.public_baseurl` instead? Or is the assumption that this is only supposed to talk back on private networks?

Reply:
well, often `config.server.public_baseurl` will point to a reverse-proxy, where we explicitly advise *against* exposing the `/_synapse/admin` namespace.

Given we're reading the homeserver.yaml, we have a reasonable intuition that synapse is running on the same host, so a hit to `localhost` is relatively likely to work.